src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -372,29 +394,42 @@ void DiscoveryDataBase::process_dispose_participant(
         }
     }
 
-    // Delete Participant entries from readers_by_topic and writers_by_topic
-    for (auto tit = readers_by_topic_.begin(); tit != readers_by_topic_.end(); ++tit)
+    // Delete Participant entries from writers_by_topic
+    for (auto tit = writers_by_topic_.begin(); tit != writers_by_topic_.end(); ++tit)
     {
-        for (auto rit = tit->second.begin(); rit != tit->second.end(); ++rit)
+        for (auto wit = tit->second.begin(); wit != tit->second.end(); ++wit)
         {
-            if(rit->guidPrefix == participant_guid.guidPrefix)
+            if (wit->guidPrefix == participant_guid.guidPrefix)
             {
-                tit->second.erase(rit);
-                --rit;
+                tit->second.erase(wit);
+                --wit;
             }
         }
+
+        if (tit->second.empty())
+        {
+            writers_by_topic_.erase(tit);
+            --tit;
+        }
     }
```

Comment:
I think this'd more optimal if we only check the topics for which we know the participant has a writer. I'll let it pass now, but I wanted to make a note somewhere of this improvement. Same goes for clearing `readers_by_topic_`